### ISSUE TITLE
Documentation suggestion: unit-level contrasts can be very meaningful

### ISSUE THREAD

**Issue by arthur-albuquerque** [vincentarelbundock/marginaleffects#448]
Hi Vincent,

In your [“Contrasts” vignette](https://vincentarelbundock.github.io/marginaleffects/articles/contrasts.html#average-contrasts), you suggest that unit-levels contrasts “…can be unwieldy and hard to interpret.”

Although I overall agree, I would like to highlight that a focus in these unit-levels contrasts has been rising in the medical literature. For example, in this [excellent blog post](https://www.fharrell.com/post/rdist/), Dr. Frank Harrell discusses possible summaries from a logistic regression model. He uses a dataset from a famous RCT, [“GUSTO-I”](https://www.nejm.org/doi/full/10.1056/NEJM199309023291001). He argues that a graph depicting all unit-level contrasts (or, as he states, “distribution of risk difference”) is much more informative than “one-number summaries”.

> Clinicians can readily see that most patients are lower risk and receive little absolute benefit form t-PA, while a minority of very high-risk patients can receive almost an absolute 0.05 risk reduction.

Hence, I suggest adding a small case study in `marginaleffects`’ documentation showing how the package can be easily used to fill this new trend in the medical literature. Of course, feel free to fully ignore this suggestion if you think it’s not in the scope.

The code below is inspired by Dr. Harrell's blog post mentioned above. I changed some details since `marginaleffects` does not support logistic regression models fitted with `rms::lrm()`. Nevertheless, I can reproduce his graphs fully with `glm()`.

----------------------------------------------------------------------------------------------

Load packages and data

``` r
library(rms)
library(data.table)
library(ggplot2)
library(marginaleffects)

load(url('https://hbiostat.org/data/gusto.rda'))
setDT(gusto)
gusto <- gusto[tx %in% c('SK', 'tPA'),
               .(day30, tx, age, Killip, sysbp, pulse, pmi, miloc, sex)]
gusto[, tx := tx[, drop=TRUE]]

gusto$tx = factor(gusto$tx, levels = c("tPA", "SK"))
```

Fit the full covariate-adjusted model.

``` r
f <- glm(day30 ~ tx + rcs(age,4) + Killip + pmin(sysbp, 120) + lsp(pulse, 50) +
         pmi + miloc + sex, family = "binomial", data=gusto)
```

## Manual approach

Distribution of absolute risk (proportion) difference

``` r
d  <- gusto

d$tx = "SK"
p1 <- predict(f, newdata = d, type = "response")

d$tx = "tPA"
p2 <- predict(f, newdata = d, type = "response")

diff = p1 - p2

data.frame(ARD = diff) |> 
  
  ggplot(aes(x = ARD)) +
  geom_histogram(color = "white", bins = 100) +
  labs(x = "SK - t-PA Risk Difference (aka, 'unit-level contrasts')") +
  theme_minimal()
```

![](https://i.imgur.com/jFt0xId.png)

# {marginaleffects} approach

``` r
comparisons(f, variables = "tx") |> 
  
  ggplot(aes(x = comparison)) +
  geom_histogram(color = "white", bins = 100) +
  labs(x = "SK - t-PA Risk Difference (aka, 'unit-level contrasts')") +
  theme_minimal()
```

![](https://i.imgur.com/fmiFuYQ.png)

``` r
sessionInfo()
#> R version 4.1.2 (2021-11-01)
#> Platform: x86_64-apple-darwin17.0 (64-bit)
#> Running under: macOS Catalina 10.15.7
#> 
#> Matrix products: default
#> BLAS:   /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRblas.0.dylib
#> LAPACK: /Library/Frameworks/R.framework/Versions/4.1/Resources/lib/libRlapack.dylib
#> 
#> locale:
#> [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
#> 
#> attached base packages:
#> [1] stats     graphics  grDevices utils     datasets  methods   base     
#> 
#> other attached packages:
#> [1] marginaleffects_0.7.0.9000 data.table_1.14.3         
#> [3] rms_6.3-0                  SparseM_1.81              
#> [5] Hmisc_4.7-0                ggplot2_3.3.6             
#> [7] Formula_1.2-4              survival_3.3-1            
#> [9] lattice_0.20-45           
#> 
#> loaded via a namespace (and not attached):
#>  [1] httr_1.4.3          splines_4.1.2       assertthat_0.2.1   
#>  [4] highr_0.9           latticeExtra_0.6-29 yaml_2.3.5         
#>  [7] pillar_1.7.0        backports_1.4.1     quantreg_5.93      
#> [10] glue_1.6.2          digest_0.6.29       RColorBrewer_1.1-3 
#> [13] checkmate_2.1.0     colorspace_2.0-3    sandwich_3.0-2     
#> [16] htmltools_0.5.2     Matrix_1.4-1        pkgconfig_2.0.3    
#> [19] purrr_0.3.4         mvtnorm_1.1-3       scales_1.2.0       
#> [22] jpeg_0.1-9          MatrixModels_0.5-0  htmlTable_2.4.0    
#> [25] tibble_3.1.7        generics_0.1.3      farver_2.1.1       
#> [28] ellipsis_0.3.2      TH.data_1.1-1       withr_2.5.0        
#> [31] nnet_7.3-17         cli_3.3.0           mime_0.12          
#> [34] magrittr_2.0.3      crayon_1.5.1        polspline_1.1.20   
#> [37] evaluate_0.15       fs_1.5.2            fansi_1.0.3        
#> [40] nlme_3.1-158        MASS_7.3-57         xml2_1.3.3         
#> [43] foreign_0.8-82      tools_4.1.2         lifecycle_1.0.1    
#> [46] multcomp_1.4-19     stringr_1.4.0       munsell_0.5.0      
#> [49] reprex_2.0.1        cluster_2.1.3       compiler_4.1.2     
#> [52] rlang_1.0.4         grid_4.1.2          rstudioapi_0.13    
#> [55] htmlwidgets_1.5.4   base64enc_0.1-3     labeling_0.4.2     
#> [58] rmarkdown_2.14      gtable_0.3.0        codetools_0.2-18   
#> [61] curl_4.3.2          DBI_1.1.3           R6_2.5.1           
#> [64] gridExtra_2.3       zoo_1.8-10          knitr_1.39         
#> [67] dplyr_1.0.9         fastmap_1.1.0       utf8_1.2.2         
#> [70] insight_0.18.0.4    stringi_1.7.6       vctrs_0.4.1        
#> [73] rpart_4.1.16        png_0.1-7           tidyselect_1.1.2   
#> [76] xfun_0.31
```

<sup>Created on 2022-08-12 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by arthur-albuquerque** [comment]
In fact, Dr. Harrell suggests that the graph above should be shown with the one below, so one could see how "unit-level contrasts" in the logistic regression vary based on the baseline risk.

> But first show the full information needed for medical decision making: the estimated risks for individual patients under both treatment alternatives. One can readily see risk magnification in the absolute risk reduction as baseline risk increases.

> Clinicians can readily see that most patients are lower risk and receive little absolute benefit form t-PA, while a minority of very high-risk patients can receive almost an absolute 0.05 risk reduction.

![image](https://user-images.githubusercontent.com/77208087/184405184-41cbf1db-d32d-4622-8012-372a9c61e73e.png)

Currently, one cannot easily plot the histogram depicted in my original post along with the figure above from `comparisons()` output (since it doesn't contain predicted outcomes in each group by row, only their difference). However, I believe this could be easily added...

In summary, I suggest:
- To depict the difference and predicted outcomes for both groups in `comparisons()` output.
- To add some discussion on how unit-level contrasts can be useful in `marginaleffects`' documentation

**Comment by ngreifer** [comment]
@arthur-albuquerque This is a really interesting application! Regarding your first bullet, you can get the (counterfactual) predicted outcomes under each treatment level using `predictions()`; see #393 (though you still have to manually merge them to get a data frame with a row for each unit and separate columns for the counterfactual predictions).

**Comment by arthur-albuquerque** [comment]
@ngreifer yes, I saw that. I believe that having a one-line command to do all this would be a nice addition to the package.

It seems that `comparisons()` applies [get_contrasts](https://github.com/vincentarelbundock/marginaleffects/blob/df0991fdd570033e79c6143946b8689d7ccd6862/R/get_contrasts.R), which uses [get_predicted](https://github.com/vincentarelbundock/marginaleffects/blob/df0991fdd570033e79c6143946b8689d7ccd6862/R/get_predict.R) under the hood. I think this workflow facilitates adding this new functionality to `comparisons`, but as I've never developed a package, so can't really say how much work that would be.

**Comment by vincentarelbundock** [comment]
Thanks for the link and discussion @arthur-albuquerque. The blog post is very cool.

The good news is that this is already super easy to do using the current development version from Github. See illustration below.

It would be fantastic to add a case study with this information. If you have time to get the ball rolling on this, I’ll be happy to review and merge. Steps would be:

1.  Copy the structure of one of the vignettes into a new file.
2.  Write the case study
3.  Add a link to the new vignette in `vignettes/toc.Rmd`
4.  Add a link to the new vignette in `vignettes/documentation.Rmd`
5.  Add an entry in .Rbuildignore to ignore the case study when it gets built on CRAN
6.  Modify the sentence in the Contrasts vignette with a link to the new case study.

Here’s a stripped down version of your example:

``` r
library(marginaleffects)
library(rms)

load(url('https://hbiostat.org/data/gusto.rda'))
gusto$tx <- factor(as.character(gusto$tx), levels = c("tPA", "SK"))

mod <- glm(
    day30 ~ tx + rcs(age, 4) + Killip + pmin(sysbp, 120) + lsp(pulse, 50) +
    pmi + miloc + sex, family = "binomial",
    data = gusto)

cmp <- comparisons(mod)
summary(cmp)
#>      Term            Contrast     Effect Std. Error z value   Pr(>|z|)
#> 1      tx            SK - tPA  0.0109247  2.779e-03   3.931 8.4768e-05
#> 2     age         (x + 1) - x  0.0039533  1.543e-04  25.622 < 2.22e-16
#> 3  Killip              II - I  0.0381063  4.200e-03   9.074 < 2.22e-16
#> 4  Killip             III - I  0.0873905  1.261e-02   6.928 4.2781e-12
#> 5  Killip              IV - I  0.1851924  2.433e-02   7.611 2.7258e-14
#> 6   sysbp         (x + 1) - x -0.0012599  8.094e-05 -15.566 < 2.22e-16
#> 7   pulse         (x + 1) - x  0.0006948  9.387e-05   7.402 1.3433e-13
#> 8     pmi            yes - no  0.0291960  3.786e-03   7.712 1.2342e-14
#> 9   miloc    Other - Inferior  0.0141308  7.451e-03   1.897    0.05789
#> 10  miloc Anterior - Inferior  0.0304963  2.917e-03  10.456 < 2.22e-16
#> 11    sex       female - male  0.0162467  3.070e-03   5.291 1.2139e-07
#> 
#> Model type:  glm 
#> Prediction type:  response

cmp <- comparisons(mod, variables = "tx")

hist(cmp$comparison,
     breaks = 100,
     main = "Distribution of unit-level treatment effects",
     xlab = "SK - tPA")
abline(v = mean(cmp$comparison), col = "red")
abline(v = median(cmp$comparison), col = "blue")
```

![](https://i.imgur.com/g4hpiS1.png)

``` r


plot(x = cmp$predicted_hi,
     y = cmp$predicted_lo,
     main = "Risk of Mortality",
     xlab = "SK",
     ylab = "t-PA")
abline(0, 1)
```

![](https://i.imgur.com/921l5r0.png)

**Comment by arthur-albuquerque** [comment]
I'm glad you liked it. Should we contact Dr. Harrell first?

**Comment by vincentarelbundock** [comment]
Hi @harrelfe !

We really like your blog post on "Avoiding One-Number Summaries of Treatment Effects for RCTs with Binary Outcomes". Thanks for sharing it!

In my last post in this thread, I show how we can replicate your analyses very easily using the `marginaleffects` package [(the code requires the Github development version).](https://vincentarelbundock.github.io/marginaleffects/#installation)

Would it be OK if we published a short case study with this code and your data on the `marginaleffects` website to illustrate the process and the intuition?

Thanks for considering!

**Comment by arthur-albuquerque** [comment]
He allowed it: https://discourse.datamethods.org/t/discussion-of-assessing-heterogeneity-of-treatment-effect-estimating-patient-specific-efficacy-and-studying-variation-in-odds-ratios-risk-ratios-and-risk-differences/1513/20

@vincentarelbundock I am no expert on collabs in GitHub. I forked your repo to my GitHub, create a new brach and pushed all your requests to that. Please let me know if that's correct.

@ngreifer it would be nice if you could also review that too. I added a few lines on marginal odds and risk ratios.

**Comment by vincentarelbundock** [comment]
@arthur-albuquerque This looks great! Thanks for taking the time. What you need to do now is open a "Pull Request". This will allow me to review it and merge it.

Normally, if you visit the main repo (not yours) on the web, you should see a yellow band warning at the top where you can click to open a pull request. If you don't see that, you can find more instructions here: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request

**Comment by arthur-albuquerque** [comment]
Done ;)
